media/open_zim.ksy

```diff
@@ -0,0 +1,380 @@
+            size: info.ptr_size * blob_count_plus_one
+            type: blobs_ptrs(blob_count)
+            if: info.compression == compression::lzma2
+          - id: blobs_ptrs_compressed_zstd
+            process: kaitai.compress.zstd
+            size: info.ptr_size * blob_count_plus_one
+            type: blobs_ptrs(blob_count)
+            if: info.compression == compression::lzma2
+          - id: blobs_ptrs_uncompressed
+            type: blobs_ptrs(blob_count)
+            if: info.compression == compression::none or info.compression == compression::default
+        instances:
+          first_blob_offset:
+            value: ptr + 1 # info_t.size
+          blobs:
+            type: blob(_index)
+            repeat: expr
+            repeat-expr: blob_count
+          first_blob_ptr_compressed:
+            pos: first_blob_offset
+            process: kaitai.compress.lzma(2)
+            size: 8 + _root.minimal_xz_lzma_size # max(8, 4) + minimal_xz_lzma_size
+            type: blob_ptr_u
+            if: info.compression == compression::lzma2
+          first_blob_ptr_uncompressed:
+            pos: first_blob_offset
+            type: blob_ptr_u
+            if: info.compression == compression::none or info.compression == compression::default
+          first_blob_ptr_u_v:
+            value: (info.compression == compression::lzma2?first_blob_ptr_compressed:first_blob_ptr_uncompressed).as<blob_ptr_u>.value
+          blob_count_plus_one:
+            value: first_blob_ptr_u_v / info.ptr_size
+          blob_count:
+            value: blob_count_plus_one - 1
+          blobs_ptrs:
+            value: |
+              (
+                info.compression == compression::lzma2
+              ?
+                blobs_ptrs_compressed_lzma2
+              :
+                (
+                  info.compression == compression::zstd
+                ?
+                  blobs_ptrs_compressed_zstd
+                :
+                  blobs_ptrs_uncompressed
+                )
+              ).ptrs
+        types:
+          blobs_ptrs:
+            params:
+              - id: count
+                type: u8
+            seq:
+              - id: ptrs
+                type: blob_ptr_u
+                repeat: eos
+            instances:
+              info:
+                value: _parent.info
+          blob:
+            params:
+              - id: idx
+                type: u8
+            instances:
+              data:
+                pos: _parent.ptr + _parent.blobs_ptrs[idx].as<blob_ptr_u>.value
+                size: _parent.blobs_ptrs[idx + 1].as<blob_ptr_u>.value - _parent.blobs_ptrs[idx].as<blob_ptr_u>.value
+          info_t:
+            #seq:
+            #  - id: compression
+            #    type: b4
+            #    enum: compression
+            #  - id: is_ptr_u8
+            #    -orig-id: extended
+            #    type: b1
+            #  - id: reserved
+            #    type: b3
```

Comment:
You can use this with [`/meta/bit-endian: le`](https://doc.kaitai.io/user_guide.html#bit-ints-le). This would probably make sense, because the whole format is little-endian as well. Also, the fields in the spec are also listed in the order from the least-significant bit (https://wiki.openzim.org/wiki/ZIM_file_format_old_namespace#Clusters):

> Field Name | Type | Offset | Length | Description
> -- | -- | -- | -- | --
> cluster information | integer | 0 | 1 | **Fourth low bits** : 0: default (no compression), 1: none (inherited from Zeno), 4: LZMA2 compressed<br><br>**Firth bits** : 0: normal (OFFSET_SIZE=4) 1: extended (OFFSET_SIZE=8)

Reply:
I prefer to keep it listed in `be` order because it is the convention KS uses for everything. I perceive `bit-endian` as a temporary solution, that IMHO was a mistake to implement that feature that way (because it causes issues for switcheable endianness). IMHO what you call `bit endianness` should be related to just endianness and to the fact of presence of bit-sized types crossing byte boundaries. I guess eventually it would be dropped and replaced to something more uniform, rather than make spec devs to code different types for be and le.